### PR TITLE
feat/MZAD-61/Auth-Module

### PR DESCRIPTION
**Pull request by beqagerliAni**

### CODE REVIEW

File: src/auth/auth.constants.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export const JwtConstants: { secret: string } = {
+  secret: 'dnck',
```
Comment (pailodze) [inline src/auth/auth.constants.ts:2]: variable name should be camelCase

File: src/auth/auth.controller.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import { User } from 'src/users/entities/users.entity';
+import { AuthService } from './auth.service';
+import { AuthDto } from './dto/auth.dto';
+import { Public } from './gurads/guard.key';
+import { LoginInterface } from './interface/login.response';
+
+@Controller('auth')
+export class AuthController {
+  constructor(private authsService: AuthService) {}
+  @Post('login')
```
Comment (pailodze) [inline src/auth/auth.controller.ts:11]: refomat this code add spaces this is not readable

File: src/auth/auth.controller.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  @Public()
+  login(@Body() authDto: AuthDto): Promise<LoginInterface> {
```
Comment (pailodze) [inline src/auth/auth.controller.ts:12]: You sure login and register has the same dtos?

File: src/auth/auth.service.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import { JwtService } from '@nestjs/jwt';
+import * as bcrypt from 'bcrypt';
+import { User } from 'src/users/entities/users.entity';
+import { UsersRepository } from 'src/users/repositories/users.repository';
+import { JwtConstants } from './auth.constants';
+import { AuthDto } from './dto/auth.dto';
+import { LoginInterface } from './interface/login.response';
+@Injectable()
+export class AuthService {
+  constructor(
```
Comment (pailodze) [inline src/auth/auth.service.ts:11]: same here add spaces the code is not readable

File: src/auth/auth.service.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    private usersRepository: UsersRepository,
+    private jwtService: JwtService,
+  ) {}
+  async register(authDto: AuthDto): Promise<User> {
+    const { email, password }: AuthDto = authDto;
+    const salt: string = await bcrypt.genSalt();
+    const hashedpassword: string = await bcrypt.hash(password, salt);
```
Comment (pailodze) [inline src/auth/auth.service.ts]: variable name should be hashedPassword

File: src/auth/auth.service.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+import { LoginInterface } from './interface/login.response';
+@Injectable()
+export class AuthService {
+  constructor(
+    private usersRepository: UsersRepository,
+    private jwtService: JwtService,
+  ) {}
+  async register(authDto: AuthDto): Promise<User> {
+    const { email, password }: AuthDto = authDto;
+    const salt: string = await bcrypt.genSalt();
```
Comment (pailodze) [inline src/auth/auth.service.ts:19]: add spaces here also

File: src/auth/auth.service.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    const user: User = await this.usersRepository.create({
+      email,
+      password: hashedpassword,
+    });
+    return user;
+  }
+  async login(authDto: AuthDto): Promise<LoginInterface> {
+    const { email, password } = authDto;
+    const user: User = await this.usersRepository.findEmail(email);
+    const chekpassword: boolean = await bcrypt.compare(password, user.password);
```
Comment (pailodze) [inline src/auth/auth.service.ts]: name this variable isPasswordCorrect = user && await bcrypt.compare(password, user.password)

File: src/auth/auth.service.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        email: string;
+        password: string;
+        role: string;
+      } = {
+        email: user.email,
+        password: user.password,
+        role: user.role,
+      };
+      return {
+        access_token: await this.jwtService.signAsync(payload, {
```
Comment (pailodze) [inline src/auth/auth.service.ts]: return accessToken not access_token

File: src/auth/enum/user.role.ts [hunk 9]
```diff
@@ -0,0 +1,4 @@
+export enum role {
```
Comment (pailodze) [inline src/auth/enum/user.role.ts]: name should be RoleEnum

File: src/auth/enum/user.role.ts [hunk 10]
```diff
@@ -0,0 +1,4 @@
...
+  admin = 'admin',
```
Comment (pailodze) [inline src/auth/enum/user.role.ts]: Enum keys should be PascalCase

File: src/auth/enum/user.role.ts (hunk 9 above)
Comment (pailodze) [inline src/auth/enum/user.role.ts]: File name should be user-roles.enum.ts

File: src/auth/gurads/auth.guard.ts [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+} from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { JwtService } from '@nestjs/jwt';
+import { Request } from 'express';
+import { JwtConstants } from 'src/auth/auth.constants';
+import { role } from '../enum/user.role';
+import { IS_PUBLIC_KEY } from './guard.key';
+import { ROLES_KEY } from './gurad.interface';
+
+interface JwtPayload {
```
Comment (pailodze) [inline src/auth/gurads/auth.guard.ts]: extract this interface to its own file and name it JwtPayloadInterface

File: src/auth/gurads/auth.guard.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  verifyAsync: boolean;
+  secret: string;
+  role: role;
+}
+
+interface RoleMetadata {
```
Comment (pailodze) [inline src/auth/gurads/auth.guard.ts]: same this name RoleMetadataInterface and extract to its own folder

File: src/auth/gurads/auth.guard.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        secret: JwtConstants.secret,
+      });
+
+      const roles: RoleMetadata =
+        this.reflector.getAllAndOverride<RoleMetadata>(ROLES_KEY, [
+          context.getHandler(),
+          context.getClass(),
+        ]);
+
+      if (roles && roles.roles && roles.roles.length) {
```
Comment (pailodze) [inline src/auth/gurads/auth.guard.ts]: const isRouteGuardedWithRoles = roles.length 
why do we need roles.roles.length?

File: src/auth/gurads/role.guard.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
+import { CanActivate, ExecutionContext, Injectable } from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { role } from '../enum/user.role';
+import { ROLES_KEY } from './gurad.interface';
+@Injectable()
+export class RolesGuard implements CanActivate {
+  constructor(private reflector: Reflector) {}
```
Comment (pailodze) [inline src/auth/gurads/role.guard.ts]: why do you need role.guard? you have written logic already in auth.guard

File: src/auth/interface/login.response.ts [hunk 15]
```diff
@@ -0,0 +1,3 @@
+export class LoginInterface {
+  access_token: string;
```
Comment (pailodze) [inline src/auth/interface/login.response.ts]: it shoudl be accessToken

File: src/users/entities/users.entity.ts [hunk 16]
```diff
@@ -14,15 +15,12 @@ export class User {
...
-  @Column()
-  lastName: string;
-
-  @Column()
-  biography: string;
-
   @Column()
   password: string;
 
+  @Column({ default: 'user' })
```
Comment (pailodze) [inline src/users/entities/users.entity.ts]: take the default value from enum

File: src/users/repositories/users.repository.ts [hunk 17]
```diff
@@ -33,4 +31,7 @@ export class UsersRepository {
   async remove(id: number): Promise<DeleteResult> {
     return await this.usersRepository.delete(id);
   }
```
Comment (pailodze) [inline src/users/repositories/users.repository.ts:35]: add space hrere

File: src/users/users.controller.ts [hunk 18]
```diff
@@ -18,6 +19,7 @@ export class UsersController {
   constructor(private readonly usersService: UsersService) {}
 
   @Post()
+  @Public()
```
Comment (pailodze) [inline src/users/users.controller.ts]: why this is public?

**Review by pailodze — CHANGES_REQUESTED** [review]

File: src/users/users.controller.ts (hunk 18 above)
Comment (beqagerliAni) [reply src/users/users.controller.ts]: gavteste prosta

File: src/auth/auth.controller.ts [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+export class AuthController {
+  constructor(private authsService: AuthService) {}
+  @Post('login')
+  login(@Body() authDto: AuthDto): Promise<LoginInterface> {
+    return this.authsService.login(authDto);
+  }
+
+  @Public()
+  @Post()
+  register(@Body() authdto: AuthDto): Promise<User> {
```
Comment (pailodze) [inline src/auth/auth.controller.ts]: argument name should be camel case

File: src/auth/auth.service.ts [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+
+  async login(authDto: AuthDto): Promise<LoginInterface> {
+    const { email, password } = authDto;
+
+    const user: User = await this.usersRepository.findEmail(email);
+
+    const isPasswordCorrect: boolean =
+      user && (await bcrypt.compare(password, user.password));
+
+    if (user && isPasswordCorrect) {
```
Comment (pailodze) [inline src/auth/auth.service.ts]: if(isPasswordCorrect)

File: src/auth/enum/user.role.ts [hunk 21]
```diff
@@ -0,0 +1,4 @@
+export enum RoleEnum {
+  ADMIN = 'ADMIN',
```
Comment (pailodze) [inline src/auth/enum/user.role.ts]: Admin = 'admin',
User = 'user'

File: src/auth/gurads/auth.guard.ts [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+        secret: jwtConstants.secret,
+      });
+
+      const roles: RoleMetadata =
+        this.reflector.getAllAndOverride<RoleMetadata>(ROLES_KEY, [
+          context.getHandler(),
+          context.getClass(),
+        ]);
+      const isRouteGuardedWithRoles: number = roles.length;
+      if (roles && roles.roles && isRouteGuardedWithRoles) {
```
Comment (pailodze) [inline src/auth/gurads/auth.guard.ts]: if(isRouteGuardedWithRoles)

File: src/auth/interface/payload.response.ts [hunk 23]
```diff
@@ -0,0 +1,7 @@
+import { RoleEnum } from '../enum/user.role';
+
+export interface JwtPayload {
+  verifyAsync: boolean;
```
Comment (pailodze) [inline src/auth/interface/payload.response.ts]: why this is here?

File: src/auth/auth.service.ts [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
```
Comment (pailodze) [inline src/auth/auth.service.ts]: you forgot this!

**Review by pailodze — CHANGES_REQUESTED** [review]

File: src/auth/enum/user.role.ts [hunk 25]
```diff
@@ -0,0 +1,4 @@
+export enum RoleEnum {
+  ADMIN = 'admin',
```
Comment (pailodze) [inline src/auth/enum/user.role.ts]: Admin
User

**Review by pailodze — CHANGES_REQUESTED** [review]

**Pull request merged**